stl/inc/system_error

```diff
@@ -483,18 +485,14 @@ public:
     }
 
     _NODISCARD virtual string message(int _Errcode) const override {
-        const unsigned long _Size = 32767;
-        string _Narrow(_Size, '\0');
-
-        const unsigned long _Val = _Winerror_message(static_cast<unsigned long>(_Errcode), &_Narrow[0], _Size);
-        if (_Val == 0) {
-            _Narrow = "unknown error";
+        const char* _Msg     = nullptr;
+        const size_t _Length = _Winerror_message2(static_cast<unsigned long>(_Errcode), &_Msg);
+        if (_Length != 0) {
+            const unique_ptr<const char[], decltype(&_Winerror_message2_free)> _Ptr(_Msg, &_Winerror_message2_free);

```

Comment:
Using a function pointer for a deleter in unique_ptr is suboptimal (it makes the unique_ptr bigger and usually prevents UTC from inlining the deleter)

Reply:
UTC is the very Microsoft-internal name for C2, the less Microsoft-internal name for MSVC's compiler backend (implemented in `c2.dll` loaded by `cl.exe`).

Reply:
Hmmm interesting: it looks like the optimizer was able to fix this case: https://gcc.godbolt.org/z/cuuDGb

I still am not a fan of this function pointer thing but now up to you :)

Reply:
![image](https://user-images.githubusercontent.com/1544943/73114548-7b3df880-3ed0-11ea-8f07-95e7a739824c.png)

Reply:
Do you have any facility similar to the previously proposed `std::monostate_function` or I should just write a one-off deleter functor here?

Reply:
Honestly if it were me I wouldn't even use unique_ptr for this: I'd write the "guard" immobile version:

```
struct _Winerror_message_caller {
size_t _Length;
const char * _Msg;
_Winerror_message_caller(unsigned long _Ec) noexcept
:_Length(_Winerror_message2(_Ec, &_Msg))
{}
_Winerror_message_caller(const _Winerror_message_caller&) = delete;
_Winerror_message_caller& operator=(const _Winerror_message_caller&) = delete;
~_Winerror_message_caller() {
_Winerror_message2_free(_Msg);
}
```

The point of unique_ptr and deleters is for when you need to pass that around with move semantics, but we don't need that here.

Reply:
This is what I ended up doing - resolved